src/react/react-decor-function.ts

```diff
@@ -0,0 +1,120 @@
+    return <P extends {}>(res: ElementArgs<P>, hook: ElementHook) => hook(componentProps, res);
+};
+
+export function decorReact<T = any>(hooks: DecorReactHooks): (comp: SFC<T>) => SFC<T> {
+    let componentProps: Attributes & Partial<any>;
+    let createArgsMap: Map<object, ElementArgs<any>> = new Map();
+    const decors: {before: Array<BeforeHook<any>>, after: Array<AfterHook<any>>} = {
+        before: [],
+        after: []
+    };
+
+    if (hooksDefined(hooks.nodes) || hooksDefined(hooks.root)) {
+        const replaceCreateElement = (args: CreateElementArgsTuple<any>): CreateElementArgsTuple<any> => {
+            componentProps = args[0]; // [0] for props in a functional react component
+            React.createElement = makeCustomCreateElement(hooks, componentProps, createArgsMap);
+            return (args as CreateElementArgsTuple<any>);
+        };
+
+        const applyRootHooks = <P extends {}>(renderResult: ReactElement<P>): ReactElement<P> => {
+            if (hooksDefined(hooks.root)) {
+                let rootElementArgs = createArgsMap.get(renderResult);
+
+                rootElementArgs = (hooks as { root: Array<ElementHook> }).root.reduce(getHooksReducer(componentProps), rootElementArgs);
+
+                if (rootElementArgs) {
+                    renderResult = cloneElement(renderResult, rootElementArgs.elementProps);
+                } else {
+                    console.warn('unable to found matching component for: ', rootElementArgs); // todo: test this?
+                }
+            }
+
+            createArgsMap.clear();
+            return renderResult;
+        };
+
+        decors.before.push(replaceCreateElement);
+        decors.after.push(applyRootHooks);
+        React.createElement = originalCreateElement as ReactCreateElement;
+    };
+
+    return decorFunction(decors) as (comp: SFC<T>) => SFC<T> ; // as SFC?
+}
+
+// create a custom react create element function that applies the given hooks
+function makeCustomCreateElement<P extends {}>(hooks: DecorReactHooks, componentProps: object, createArgsMap: Map<object, ElementArgs<P>>): typeof React.createElement {
+    if (hooksDefined(hooks.nodes) || hooksDefined(hooks.root)) {
+        let createElementArgsObject: ElementArgs<P>;
+
+        const applyHooksOnArguments = (createElementArgsTuple: CreateElementArgsTuple<P>): CreateElementArgsTuple<P> => {
+            createElementArgsObject = translateArgumentsToObject(createElementArgsTuple);
+            if (hooksDefined(hooks.nodes)) {
+                createElementArgsObject = (hooks as { nodes: Array<ElementHook> }).nodes.reduce(getHooksReducer(componentProps), createElementArgsObject);
+                return createElementArgsTuple = translateObjectToArguments(createElementArgsObject);
+            }
+
+            return createElementArgsTuple;
+        };
+
+        const saveCreateElementArguments = (createElementResult: ReactNode): ReactNode => {
```

Comment:
`React.createElement` does not return `ReactNode`. it's the wrong type to use and it's the reason for force-casting it into the map.